A slightly more sophisticated Makefile

Fat-Zer opened the pull request:
- support for override of the compiler and utils
- support for append/override CFLAGS
- a target to update man pages
- install target
- better handling of conditional flags, particularly:
  - disable some warnings on non-gcc compilers
  - handle GNU's tuples like `x86-64-pc-linux-gnu` and `aarch-unknowv-linux-gnu`
- organize CFLAGS by sorting them into categories

This pull request changes...

## Changes

## Mandatory Checklist

- [X] Legal agreements accepted. By contributing to this organisation, you acknowledge you have read, understood, and agree to be bound by these these agreements:

[Terms of Service](https://www.whonix.org/wiki/Terms_of_Service), [Privacy Policy](https://www.whonix.org/wiki/Privacy_Policy), [Cookie Policy](https://www.whonix.org/wiki/Cookie_Policy), [E-Sign Consent](https://www.whonix.org/wiki/E-Sign_Consent), [DMCA](https://www.whonix.org/wiki/DMCA), [Imprint](https://www.whonix.org/wiki/Imprint)

## Optional Checklist
The following items are optional but might be requested in certain cases.

- [X] I have tested it locally
- [ ] I have reviewed and updated any documentation if relevant
- [ ] I am providing new code and test(s) for it

Fixes #

ArrayBolt3 reviewed (changes requested):
Overall I really like the changes here, it makes things much more flexible. Worthy of note, some changes are going to need to be made to the Debian packaging in order to support this. Specifically `debian/kloak.install` will need to be deleted entirely, and `debian/rules` will need some tweaks to set the prefix correctly and not use `override_dh_installman` anymore. This will probably also break the RPM spec file that was contributed recently, so I think @siliconwaffle should take a look at this too since they're the ones who contributed that part of things.

siliconwaffle commented:
Thanks for @ me @ArrayBolt3, you're correct that this is messy for Fedora packaging. I think the apparmor install should be excluded or separated from the install target, because not all systems have apparmor, it's mainly a Debian thing(and I think Suse and a couple other distros support it, but not Fedora). If it stays then it means I have to either patch the makefile, delete from the buildroot, or just not use the install target for the makefile at all, it's fine just a little messy for systems without apparmor.

As for my input:
1 - This line is unnecessary, as the other install commands after will automatically create the directories that they need.
`$(INSTALL) -d -m 755 $(addprefix $(DESTDIR), $(sbindir) $(mandir) $(udev_rules_dir) $(udev_rules_dir) $(apparmor_dir) $(systemd_dir))`

2 - For the variables defined at the top, maybe I'm just ignorant but it's a little weird, I've never seen it before. I thought these were usually defined with a configure script? Plus /usr/local is not a proper prefix, /usr/local is meant for things that aren't packaged and are instead installed by the individual user or sysadmin. /usr is usually the prefix for things that are packaged and distributed. It'd probably be easier for me to keep the .spec as is than using this install target as is. With rpms, the %configure macro usually defines the prefix, bindir, sbindir, etc. so without a configure script this doesn't quite work at least from what I understand, but I don't know much about makefiles outside of looking a few for software that I package, so take what I say with a grain of salt.

With a few edits this is a good pr I think. I wanted to add an install target but haven't gotten around to it, so this pr is appreciated.

siliconwaffle commented on Makefile:
```diff
@@ -1,6 +1,23 @@
+prefix         ?= /usr/local
+sbindir        ?= $(prefix)/sbin
+datadir        ?= $(prefix)/share
+mandir         ?= $(datadir)/man
```
Yes they belong in $(mandir)/man8/, but I don't think he should change the mandir variable, instead they should install to  $(mandir)/man8/ instead of changing the mandir variable.

ArrayBolt3 replied:
That makes sense to me, I'll amend my suggestion.

Fat-Zer replied:
Nah... the expected value for mandir is `$(datadir)/man`. This should rather be better handled in install rules. My mistake that I've overlooked it initiall. Thanks for heads-up.

fixed

ArrayBolt3 commented:
It is doing some of the same jobs as a configure script since we don't have a configure script in Kloak (and personally I'd like to avoid autotools if at all possible). The apparmor thing is a good point though, since Fedora typically uses SELinux.

I'm not sure what the best approach is here. Would it be better for there to be a separate target for installing apparmor, similar to how the Linux Kernel uses separate targets for installing the kernel, modules, and headers?

Also, if this is getting messy for distro packaging, I'm starting to doubt if we actually want to integrate this PR. I originally thought it would make distro packaging *easier*, but it sounds like that may not be the case. If it is helpful, then great, but otherwise, maybe it's better to just stick with the "dumb" makefile?

siliconwaffle commented:
It's not that Fedora _typically_ uses selinux, Fedora _only_ uses selinux, to get apparmor support in Fedora or derivatives, you have to compile your own kernel and install apparmor yourself and install your own apparmor profiles too since no Fedora packages include it.

I think it's okay to add the install target, with a couple fixes at least it's really good. Regardless, even if `make install` is problematic for non-Debian distros, it's relatively easy to just not use it or patch it.

Fat-Zer commented on Makefile:
```diff
@@ -1,6 +1,23 @@
+
+udev_rules_dir ?= /lib/udev/rules.d
+apparmor_dir   ?= /etc/apparmor.d/
+systemd_dir    ?= $(prefix)/lib/systemd/system
```
ok, let it be so...

I've seen it too before, and at that time is sounded more consistent to put it into the prefix anyway. But now as I'm considering the case where a user has custom prefix there is no way to tell systemd to actually use it. So a default to `/usr` makes more sense.

Fat-Zer commented on Makefile:
```diff
@@ -13,34 +30,57 @@ TARGETARCH=$(shell gcc -dumpmachine)
+	-fno-strict-overflow -fno-strict-aliasing -fsanitize=undefined
+
+ifeq (yes,$(patsubst x86_64%-linux-gnu,yes,$(TARGETARCH)))
+FORTIFY_CFLAGS += -fcf-protection=full # only supporudev_rules_dirted on x86_64
```
fixed

Fat-Zer commented on Makefile:
```diff
@@ -13,34 +30,57 @@ TARGETARCH=$(shell gcc -dumpmachine)
 
 kloak : src/main.c src/keycodes.c src/keycodes.h
-	gcc -g src/main.c src/keycodes.c -o kloak -lm $(shell pkg-config --cflags --libs libevdev) $(shell pkg-config --cflags --libs libsodium) $(CPPFLAGS) $(CFLAGS) $(LDFLAGS)
+	$(CC) src/main.c src/keycodes.c -o kloak -lm $(shell $(PKG_CONFIG) --cflags --libs libevdev) $(shell $(PKG_CONFIG) --cflags --libs libsodium) $(CFLAGS) $(LDFLAGS)
```
my mistake. I actually wasn't aware it's dedicated for include paths, and thought it's alternative for `CXXFLAGS`.

fixed

ArrayBolt3 replied:
:+1:  (just fyi, since this initially confused me too, CPP in this context is "C preprocessor", not "C++", which is why it's relevant even in an application with no C++ code. I really *really* wish whoever came up with the acronyms had come up with better acronyms, but, they didn't :-/)

Fat-Zer replied:
ok, that's getting more embracing: I've just now connected the dots that `CPPFLAGS` are flags for `cpp` (as in executable name for C preprocessor) =))

Fat-Zer commented on Makefile:
```diff
@@ -13,34 +30,57 @@ TARGETARCH=$(shell gcc -dumpmachine)
+man : $(MANPAGES)
+
+auto-generated-man-pages/% : man/%.ronn
+	ronn -r $< --manual="kloak Manual" --organization="kloak" --pipe >$@
```
done

PS: AFAIR if a file is provided on stdin rather than as an argument `-r` and `--pipe` are implied, so it was the same thing.

Fat-Zer commented on Makefile:
```diff
@@ -13,34 +30,57 @@ TARGETARCH=$(shell gcc -dumpmachine)
 	rm -f kloak eventcap
+
+install : all lib/udev/rules.d/95-kloak.rules etc/apparmor.d/usr.sbin.kloak  usr/lib/systemd/system/kloak.service $(MANPAGES)
+	$(INSTALL) -d -m 755 $(addprefix $(DESTDIR), $(sbindir) $(mandir) $(udev_rules_dir) $(udev_rules_dir) $(apparmor_dir) $(systemd_dir))
```
fixed

Fat-Zer commented on Makefile:
```diff
@@ -13,34 +30,57 @@ TARGETARCH=$(shell gcc -dumpmachine)
+	-Werror=incompatible-pointer-types
+
+ifneq (,$(findstring gcc,$(CC_VERSION)))
+WARN_CFLAGS += -Wtrampolines -Wbidi-chars=any  # clang as for 18.1.8 doesn't support this warnings
```
Sounds too hacky for my taste... Since as for now the only other compiler that's in the picture is clang, let's just disable the warnings for it... If anybody will need to support other compilers it would be easier to switch to a build system from this century like cmake.

done

Fat-Zer commented:
> Worthy of note, some changes are going to need to be made to the Debian packaging in order to support this. Specifically debian/kloak.install will need to be deleted entirely, and debian/rules will need some tweaks to set the prefix correctly and not use override_dh_installman anymore. This will probably also break the RPM spec file that was contributed recently, so I think @siliconwaffle should take a look at this too since they're the one who contributed that part of things.

Not necessary, I've tried to be non-disruptive enough so all old packaging could keep working as-is. `install` is more of a hint for new packagers.

> 1 - This line is unnecessary, as the other install commands after will automatically create the directories that they need.

No, `install` doesn't do that. There is a `-D` switch, but it'd require to add target filename. And IMHO it's less error-prone to just pre-create the directories.

> For the variables defined at the top, maybe I'm just ignorant but it's a little weird, I've never seen it before. I thought these were usually defined with a configure script?

Many packages that use bare Makefiles do it:
- for CC/CFLAGS/PKGCONFIG you usually export them to the environment (and usually building systems do that automagicaly or there is an easy shortcut for that)
  Note that `CFLAGS=-O1 make` will override only "-O2 -g" in the Makefile, while `make CFLAGS="-O1"` will override all of them.
- for `prefix` etc — you usually pass them to `make` as arguments, more often just during the install phase: `make DESTDIR=/path/to/image prefix=/usr install`, but IMHO it's safer to pass them during both build and install if a package will start to embed those into sources.

  There is no reason not to pass them through env, but I've seen it to be more common, as well as use of lowercase variables for the `prefix` etc.

> Plus /usr/local is not a proper prefix, /usr/local is meant for things that aren't packaged and are instead installed by the individual user or sysadmin. /usr is usually the prefix for things that are packaged and distributed. It'd probably be easier for me to keep the .spec as is than using this install target as is. With rpms, the %configure macro usually defines the prefix, bindir, sbindir, etc. so without a configure script this doesn't quite work at least from what I understand, but I don't know much about makefiles outside of looking a few for software that I package, so take what I say with a grain of salt.

Yes, it's a common convention that the default prefix is `/usr/local`, that's targeted for poor souls who decide to run `make install` on their own (also known as junior sysadmins). In such cases packagers for distributions are expected to reset it explicitly to an appropriate value like: `make prefix=/usr install`. The same convention is true for autotools and cmake, but the prefix reset is usually hidden in the guts of packaging system.

On the other hand systemd/apparmor are hard-coding /usr/sbin/kloak. So it mat be actually makes sense to use /usr here.

> Thanks for @ me @ArrayBolt3, you're correct that this is messy for Fedora packaging. I think the apparmor install should be excluded or separated from the install target, because not all systems have apparmor, it's mainly a Debian thing(and I think Suse and a couple other distros support it, but not Fedora). If it stays then it means I have to either patch the makefile, delete from the buildroot, or just not use the install target for the makefile at all, it's fine just a little messy for systems without apparmor.

Not installing it has another problem: It's much easier for a packager to notice that there is a file installed that is not needed than to notice that there is a useful file that buried somewhere in the distro. So it likely might be unnoticed. 

Though I won't insist on this one. In both cases we can't get "one size fits all" approach: it's either Debian's packagers who will have to add another file or Fedora's who will have to remove it during installation.

----------------

So, with all sheets of text being typed, I suppose currently open questions are:

- What should be the default prefix? `/usr` or `/usr/local`?
- Install `etc/apparmor.d/usr.sbin.kloak` or not?
- Enabling `-Wtrampolines -Wbidi-chars=any` for everybody except clang instead of enabling just for gcc, ok?

ArrayBolt3 commented:
* You have a good point about systemd and AppArmor being hardcoded to `/usr/sbin/kloak`. systemd's hardcoding is (I believe) intentional, to avoid loading the wrong executable, and AppArmor's hardcoding isn't really avoidable, so I think sticking with `/usr` as the default prefix is better.
* I'm not really comfortable with leaving the AppArmor profile "floating". Distros who package it can deal with it easily enough, but users who `make install` won't get it otherwise. Unless @siliconwaffle has strong objections, I'd prefer to just install it and have the Fedora packaging remove it. It won't cause harm on AppArmor-less systems that a user `make install`s is into, distros without AppArmor can remove it, and it helps keep the user more secure on those distros that do have AppArmor support.
* The solution with `-Wtrampolines` and `-Wbidi-chars=any` sounds great to me.

I haven't re-reviewed this yet, this is just my answers to the three questions. I'll look through this again very soon.

Fat-Zer commented:
> You have a good point about systemd and AppArmor being hardcoded to /usr/sbin/kloak. systemd's hardcoding is (I believe) intentional, to avoid loading the wrong executable, and AppArmor's hardcoding isn't really avoidable, so I think sticking with /usr as the default prefix is better.

Ok, I've changed default `prefix` to `/usr` and added a note for future generations.

siliconwaffle commented:
>     * I'm not really comfortable with leaving the AppArmor profile "floating". Distros who package it can deal with it easily enough, but users who `make install` won't get it otherwise. Unless @siliconwaffle has strong objections, I'd prefer to just install it and have the Fedora packaging remove it. It won't cause harm on AppArmor-less systems that a user `make install`s is into, distros without AppArmor can remove it, and it helps keep the user more secure on those distros that do have AppArmor support.

Fair enough, sounds good to me.

ArrayBolt3 commented:
This is looking quite good! A few minor gripes remaining, mostly just typo fixes:

* `# NOTE: system.d unit and apparmor profile are hardcoded to use`: `system.d` should change to `The systemd`, so it reads `# NOTE: The systemd unit and apparmor profile are hardcoded to use
* `So if you change this default install paths, you will`: `this` should change to `the`, so it reads `So if you change the default install paths, you will`
* `# only supporued on x86_64` should change to `# only supported on x86_64` (`supporued` vs `supported`).
* `ronn --manual="kloak Manual" --organization="kloak" <$<  >$@` - there's two spaces betweeen `<$<` and `>$@` when there should only be one.
* eventcap is still missing its `$(CPPFLAGS)`.

Otherwise, this looks good. Thanks for putting up with me being nitpicky. :)

Fat-Zer commented:
all fixed

> Otherwise, this looks good. Thanks for putting up with me being nitpicky. :)

don't sweat it, I'm the same ;)

ArrayBolt3 reviewed (approved):
Looks good to me!
